plugins/storage/volume/storpool/src/main/java/org/apache/cloudstack/storage/collector/StorPoolAbandonObjectsCollector.java

```diff
@@ -322,4 +340,72 @@ private Map<String,String> getStorPoolNamesAndCsTag(JsonArray arr) {
         }
         return map;
     }
+
+    class StorPoolSnapshotRecoveryCheck extends ManagedContextRunnable {
+
+        @Override
+        protected void runInContext() {
+            List<StoragePoolVO> spPools = storagePoolDao.findPoolsByProvider(StorPoolUtil.SP_PROVIDER_NAME);
+            if (CollectionUtils.isEmpty(spPools)) {
+                return;
+            }
+            List<SnapshotDetailsVO> snapshotDetails = snapshotDetailsDao.findDetails(StorPoolUtil.SP_RECOVERED_SNAPSHOT);
+            if (CollectionUtils.isEmpty(snapshotDetails)) {
+                return;
+            }
+            Map<Long, StoragePoolVO> onePoolforZone = new HashMap<>();
+            for (StoragePoolVO storagePoolVO : spPools) {
+                onePoolforZone.put(storagePoolVO.getDataCenterId(), storagePoolVO);
+            }
+            List<Long> recoveredSnapshots = new ArrayList<>();
+            for (StoragePoolVO storagePool : onePoolforZone.values()) {
+                try {
+                    logger.debug(String.format("Checking StorPool recovered snapshots for zone [%s]",
+                            storagePool.getDataCenterId()));
+                    SpConnectionDesc conn = StorPoolUtil.getSpConnection(storagePool.getUuid(),
+                            storagePool.getId(), storagePoolDetailsDao, storagePoolDao);
+                    JsonArray arr = StorPoolUtil.snapshotsList(conn);
+                    List<String> snapshots = snapshotsForRcovery(arr);
+                    if (snapshots.isEmpty()) {
+                        continue;
+                    }
+                    for (SnapshotDetailsVO snapshot : snapshotDetails) {
+                        String name = snapshot.getValue().split(";")[0];
```

Comment:
make sure the snapshot detail after `.split()` has length >= 2 & then get name and location from it (chance of Index Out of Bound error).